core/finality-grandpa/src/communication/gossip.rs

```diff
@@ -980,6 +989,69 @@ impl<Block: BlockT> Inner<Block> {
 
 		(true, report)
 	}
+
+    /// The initial logic for filtering messages follows the given state
+    /// transitions:
+    ///
+    /// - State 0: not allowed to anyone (only if our local node is not an authority)
+    /// - State 1: allowed to random `sqrt(authorities)`
+    /// - State 2: allowed to all authorities
+    /// - State 3: allowed to random `sqrt(non-authorities)`
+    /// - State 4: allowed to all non-authorities
+    ///
+    /// Transitions will be triggered on repropagation attempts by the
+    /// underlying gossip layer, which should happen every 30 seconds.
+    fn message_allowed<N>(&self, peer: &PeerInfo<N>, mut previous_attempts: usize) -> bool {
+        const MIN_AUTHORITIES: usize = 5;
+
+        if !self.config.is_authority && previous_attempts == 0 {
+            // non-authority nodes don't gossip any messages right away. we
+            // assume that authorities (and sentries) are strongly connected, so
+            // it should be unnecessary for non-authorities to gossip all
+            // messages right away.
+            return false;
+        }
+
+        if !self.config.is_authority {
+            // since the node is not an authority we skipped the initial attempt
+            // to gossip the message, therefore we decrement `previous_attempts`
+            // so that the state machine below works the same way it does for
+            // authority nodes.
+            previous_attempts -= 1;
+        }
+
+        if peer.roles.is_authority() {
+            let authorities = self.peers.authorities();
+
+            // the target node is an authority, on the first attempt we start by
+            // sending the message to only `sqrt(authorities)` (if we're
+            // connected to at least `MIN_AUTHORITIES`).
+            if previous_attempts == 0 && authorities > MIN_AUTHORITIES {
+                let authorities = authorities as f64;
+                let p = (authorities.sqrt()).max(MIN_AUTHORITIES as f64) / authorities;
+                rand::thread_rng().gen_bool(p)
```

Comment:
This is not exactly same as selecting exactly sqrt random peers and sending to them. I think it is still fine here though.

Reply:
Yeah, unfortunately there's no easy way to do this with the current interface, should be doable after refactoring consensus gossip.